spot-contracts/contracts/RolloverVault.sol

```diff
@@ -518,13 +491,162 @@ contract RolloverVault is
+
+        // The mint fees are settled by simply minting fewer vault notes.
+        notes = notes.mulDiv(FEE_ONE_PERC - feePerc, FEE_ONE_PERC);
+
+        return notes;
+    }
+
+    /// @inheritdoc IVault
+    function computeRedemptionAmts(uint256 perpAmtBurnt) public returns (IVault.TokenAmount[] memory) {
+        //-----------------------------------------------------------------------------
+        uint256 feePerc = feePolicy.vaultBurnFeePerc();
+        //-----------------------------------------------------------------------------
+
+        uint256 totalSupply_ = totalSupply();
+        uint8 deployedCount_ = uint8(_deployed.length());
+        uint8 assetCount_ = 1 + deployedCount_;
+
+        // aggregating vault assets to be redeemed
+        IVault.TokenAmount[] memory redemptions = new IVault.TokenAmount[](assetCount_);
+
+        // underlying share to be redeemed
+        redemptions[0].token = underlying;
+        redemptions[0].amount = redemptions[0].token.balanceOf(address(this)).mulDiv(perpAmtBurnt, totalSupply_);
+        redemptions[0].amount = redemptions[0].amount.mulDiv(FEE_ONE_PERC - feePerc, FEE_ONE_PERC);
+
+        for (uint8 i = 0; i < deployedCount_; i++) {
+            // tranche token share to be redeemed
+            redemptions[i + 1].token = IERC20Upgradeable(_deployed.at(i));
+            redemptions[i + 1].amount = redemptions[i + 1].token.balanceOf(address(this)).mulDiv(
+                perpAmtBurnt,
+                totalSupply_
+            );
+            redemptions[i + 1].amount = redemptions[i + 1].amount.mulDiv(FEE_ONE_PERC - feePerc, FEE_ONE_PERC);
+        }
+
+        return redemptions;
+    }
+
+    /// @notice Computes the amount of perp tokens that are returned when user swaps a given number of underlying tokens.
+    /// @param underlyingAmtIn The number of underlying tokens the user swaps in.
+    /// @return perpAmtOut The number of perp tokens returned to the user.
+    /// @return perpFeeAmtToBurn The amount of perp tokens to be paid to the perp contract as mint fees.
+    /// @return s The pre-swap perp and vault subscription state.
+    function computeUnderlyingToPerpSwapAmt(uint256 underlyingAmtIn)
+        public
+        returns (
+            uint256,
+            uint256,
+            IFeePolicy.SubscriptionParams memory
+        )
+    {
+        // Compute equal value perps to swap out to the user
+        IFeePolicy.SubscriptionParams memory s = _querySubscriptionState();
+        uint256 perpAmtOut = underlyingAmtIn.mulDiv(perp.totalSupply(), s.perpTVL);
+
+        //-----------------------------------------------------------------------------
+        // When user swaps underlying for vault's perps -> perps are minted by the vault
+        // We thus compute fees based on the post-mint subscription state.
+        IFeePolicy.SubscriptionParams memory postSwapState;
+        postSwapState.perpTVL = s.perpTVL + underlyingAmtIn;
+        postSwapState.vaultTVL = s.vaultTVL;
```

Comment:
Value is conserved within the vault and NOT the perp+vault system. Minting results in a net increase into the combined system. The perp tvl will increase (because of minting new perps). The vault TVL will remain the same.